chainer/sequential.py

```diff
@@ -0,0 +1,393 @@
+                x = layer(x)
+        return x
+
+    def __reduce__(self):
+        n_lambda = 0
+        for layer in self._layers:
+            if callable(layer) and hasattr(layer, '__name__') \
+                    and layer.__name__ == '<lambda>':
+                n_lambda += 1
+
+        if n_lambda > 0:
+            raise ValueError(
+                'This Sequential object has at least one lambda function as '
+                'its component. Lambda function can not be pickled, so please '
+                'consider to use functools.partial instead of the lambda '
+                'function or use "dill", which is an external package that '
+                'enables pickling an object including lambda functions intead '
+                'of built-in pickle.')
+        return super(Sequential, self).__reduce__()
+
+    def __str__(self):
+        ret = ''
+        for i, layer in enumerate(self):
+            if isinstance(layer, Sequential):
+                name = layer.__class__.__name__
+                name += '\twhich has {} layers'.format(len(layer))
+            elif isinstance(layer, link.Chain):
+                name = layer.__class__.__name__
+                name += '\tThe structure behind a Chain is determined at '
+                name += 'runtime.'
+            elif isinstance(layer, link.ChainList):
+                name = layer.__class__.__name__
+                name += '\tThe structure behind a ChainList is determined at '
+                name += 'runtime.'
+            elif isinstance(layer, link.Link):
+                name = layer.__class__.__name__
+                param_info = '\t'
+                for param in sorted(layer.params(), key=lambda p: p.name):
+                    param_info += param.name
+                    if param._data[0] is not None:
+                        param_info += str(param._data[0].shape)
+                    else:
+                        param_info += '(None)'
+                    param_info += '\t'
+                name = name + param_info
+            elif isinstance(layer, function.Function):
+                name = layer.__class__.__name__
+            elif isinstance(layer, functools.partial):
+                name = repr(layer)
+            elif layer.__name__ == '<lambda>':
+                name = inspect.getsource(layer).strip()
+            else:
+                name = layer.__name__
+            ret += '{}\t{}\n'.format(i, name)
+        return ret
+
+    def append(self, layer):
+        self.insert(len(self), layer)
+
+    def extend(self, sequential):
+        for layer in sequential:
+            self.append(layer)
+
+    def insert(self, i, layer):
+        if not callable(layer):
+            raise ValueError(
+                'All elements of the argment should be callable. But '
+                'given {} is not callable.'.format(layer))
+
+        self._layers.insert(i, layer)
+        if isinstance(layer, link.Link):
+            self._children.insert(i, layer)
```

Comment:
Is it correct to use the same index between `self._layers` and `self._children`?

Reply:
Thank you for pointing out this problem. If the given `layer` is a Link, the index used to insert it into `self._children` should be `j + 1` where `j` is the index of the last `link.Link` object before the `i`-th element in `self._layers`. So I fixed this method to perform this procedure.